main/src/ca/uwaterloo/flix/language/phase/optimizer/RecursionRewriter.scala

```diff
@@ -0,0 +1,454 @@
+  *       def lastMapLoop(l1) = match l1 {
+  *           case Nil      => None
+  *           case x :: Nil => Some(f(x))
+  *           case _ :: xs  => lastMapLoop(xs)
+  *       };
+  *       lastMapLoop(l)
+  * }}}
+  */
+object RecursionRewriter {
+
+  def run(root: MonoAst.Root)(implicit flix: Flix): MonoAst.Root = flix.phase("RecursionRewriter") {
+    val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    root.copy(defs = newDefs)
+  }
+
+  private def visitDef(defn: MonoAst.Def)(implicit flix: Flix): MonoAst.Def = {
+    implicit val ctx: LocalContext = LocalContext.mk()
+    if (isRewritable(defn))
+      rewriteDefn(defn)
+    else
+      defn
+  }
+
+  private def visitExp(exp0: MonoAst.Expr, tailPos: TailPosition)(implicit sym0: Symbol.DefnSym, fparams: List[MonoAst.FormalParam], ctx: LocalContext, flix: Flix): TailPosition = exp0 match {
+    case Expr.Cst(_, _, _) =>
+      tailPos
+
+    case Expr.Var(_, _, _) =>
+      tailPos
+
+    case Expr.Lambda(_, exp, _, _) =>
+      visitExp(exp, tailPos = TailPosition.NonTail)
+
+    case Expr.ApplyAtomic(_, exps, _, _, _) =>
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.ApplyClo(exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos)
+      val t2 = visitExp(exp2, tailPos = TailPosition.NonTail)
+      TailPosition.combine(t1, t2)
+
+    case Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc) =>
+      // Check for recursion
+      if (sym == sym0) {
+        if (tailPos == TailPosition.NonTail) {
+          return tailPos
+        }
+        ctx.selfTailCalls.addOne(Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc))
+      }
+      // Check alive parameters
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.ApplyLocalDef(_, exps, _, _, _) =>
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.Let(_, exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.LocalDef(_, _, exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.Scope(_, _, exp, _, _, _) =>
+      visitExp(exp, tailPos)
+
+    case Expr.IfThenElse(exp1, exp2, exp3, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      val t3 = visitExp(exp3, tailPos)
+      TailPosition.combine(TailPosition.combine(t1, t2), t3)
+
+    case Expr.Stm(exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
```

Comment:
See my comment at the end. This kind of abstraction is non-nonsensical because if exp1 is not in tail position how can t1 be? And if t1 cannot be then why call combine? All these questions arise because we are trying to be too smart.

Reply:
(Comment at the top actually)